### DISCUSSION TITLE
CI/CD Setup

### DISCUSSION THREAD

**Discussion by fdegir** [Nordix/bomres#4] [Ideas]
## Initial CI/CD Setup for BomRes Repository

We could start setting up jobs oriented toward GitHub PR Workflow and then look into setting up periodic and manual jobs depending on the needs.

Here are few suggestions.

### PR Gating

These jobs get triggered when a new PR is proposed to the repository, running

- static analysis
- container image build
- auto-doc generation

and if any of these jobs fail, PR gets blocked from merge.
Additional jobs could be setup to run unit tests and functional tests depending on their availability.

### Post Merge

These jobs get triggered when an existing PR gets accepted and merged to main development branch, running

- container image build, tagging the built images with dev tag, and pushing them to Nordix Harbor Container Image Registry
- deployment of dev versions of services against target/staging environment and running health checks
- auto-doc generation
- running FOSS Introduction

Please note that some of the jobs/stages such as build/tag of the image with dev tag and running basic tests against the deployment could be moved to PR Gating as well depending on how much time it takes to run such things.

### Periodic

These jobs get triggered on a periodic basis (e.g., nightly), running

- container image build, tagging the built images with latest tag, and pushing them to Nordix Harbor Container Image Registry
- deployment of latest versions of services against production environments and ensuring services behave as expected by running smoke tests

### Release

These jobs are triggered manually to start with, running release activities such as tagging the git repo using semantic versioning, building container images and tagging them according to git tag, and publishing them under releases of the repo.

## Technologies to Use for CI/CD

Starting with GitHub Actions would be easy way to set up first two jobs, PR Gating and Post Merge, however other jobs may require access to certain non-GitHub systems and probably better to run them elsewhere. But these things could be decided once we set the basics up.

**Comment by hans-lammda** [comment]
Prior to the contribution we added support  to run the services as tools. If we want to continue dual use, the CI/CD should also include some support for deployment to Kubernetes, including tests.

**Comment by hans-lammda** [comment]
I found some example indicating that Alpine use gitlab for building packages. 

https://git.alpinelinux.org/aports/tree/.gitlab-ci.yml

**Comment by hans-lammda** [comment]
Currently Bomresolver is designed for rootless docker, we should consider podman. 

buildah bud -t bom_resolver  Containerfile .
